Add LM Optimism L1 to L2 bridge interface

ogtownsend opened the pull request:
This PR adds the Optimism bridge interface for L1 to L2 transfers. Some important things to note:
- The Optimism L1 bridge adapter increments a nonce with each transfer and each time `sendERC20` is called. This nonce is included in multiple events throughout the transfer's lifecycle and is how we "match" and link events together to confirm they've reached the various states.
- The `GetTransfers()` function makes use of this nonce and implements the matching logic to link these 3 events together and partition transfers into categories:
  - L1's LiquidityManager `LiquidityTransferred.bridgeReturnData`, which should equal 
  - OP's StandardBridge event `ERC20BridgeFinalized.extraData` which should equal
  - L2's LiquidityManager `LiquidityTransferred.bridgeSpecificData`
- See the `TODO: Applying the 1.2x gas...` comment on how we should think about adjusting the gas limit appropriately to account for how OP's native bridge dynamically burns gas

ogtownsend committed:
Add optimism L1 to L2 bridge interface

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/arb/l1_to_l2.go:
```diff
@@ -423,6 +425,9 @@ func partitionTransfers(
 		}
 	} else if len(sentLogs) < len(effectiveDepositFinalized) {
 		// more finalized than have been sent - should be impossible
+		// TODO: what if a rebalance is triggered at T=0, the sent log is emitted at t=1
+		// and the DepositFinalized log is emitted at T=2, and our query goes back to T=2. There will be
+		// 1 DepositFinalized log and 0 sent logs. Maybe drop this condition and just do the matching
```
cc @makramkd , we chatted about this in one of our meetings, I'll probably tackle this when I work on the arbitrum side of things

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/common/common.go:
```diff
@@ -0,0 +1,62 @@
+const (
+	DurationMonth = 720 * time.Hour
+
+	// TODO: these index values might need to be updated when Ryan makes changes to the LM contract and event fields
+	// LiquidityTransferredToChainSelectorTopicIndex is the index of the topic in the LiquidityTransferred event
+	// that contains the "to" chain selector.
```
FYI @RyanRHall , these are the indices that might change once we update the contract events and may impact the log poller queries

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/opstack/common.go:
```diff
@@ -0,0 +1,42 @@
+	ERC20BridgeFinalizedTopic = optimism_standard_bridge.OptimismStandardBridgeERC20BridgeFinalized{}.Topic()
+
+	// Optimism contract addresses: https://docs.optimism.io/chain/addresses
+	// TODO: confirm that we're able to use these L1 proxy addresses
```
@makramkd or @RyanRHall  do you know how often these proxy addresses change? Can we use these hardcoded proxy addresses in log poller queries reliably, or do they need to be fetched programmatically?

ogtownsend replied:
LM-167

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/opstack/l1_to_l2.go:
```diff
@@ -0,0 +1,564 @@
+	l1FilterName := fmt.Sprintf("OptimismL1ToL2Bridge-L1-LiquidityManager:%s-Local:%s-Remote:%s",
+		l1LiquidityManagerAddress.String(), localChain.Name, remoteChain.Name)
+
+	// TODO: FIXME pass valid context
```
@amirylm what's the correct way to pass a context down into these bridge interfaces? I see this `TODO FIXME` inside the arbitrum bridges as well

amirylm replied:
We should accept a context arg in this function, that should be passed all the way from `NewBridge()` which currently doesn't accept a context but it should.

You can also create a ticket for fixing contexts cross the liquidity manager instead of fixing it in this PR

ogtownsend replied:
Created LM-116

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/opstack/l1_to_l2.go:
```diff
@@ -0,0 +1,564 @@
+	ctx context.Context,
+	transfer models.Transfer,
+) ([]byte, *big.Int, error) {
+	// TODO: maybe add check if this is a native transfer or ERC20 transfer
```
we're only transferring ERC20s (WETH) for now right?

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/opstack/l1_to_l2.go:
```diff
@@ -0,0 +1,564 @@
+
+	// Scale gas limit by recommended 20% buffer to account for gas burned for L2 txn:
+	// https://docs.optimism.io/builders/app-developers/bridging/messaging#fees-for-sending-data-between-l1-and-l2
+	// TODO: Applying the 1.2x gas limit bump here to the fee won't really have an effect on the actual gas burned by
+	//   the OP bridge since the gas units are hardcoded to 1e6 in services/relay/evm/liquidity_manager.go. We should
+	//   instead consider a better way to dynamically bump the gas used by the transmitter, or just hardcode an even
+	//   higher gas limit in the transmitter. The OP team has confirmed that only gas up to the "market rate" will be
+	//   burned, not all gas remaining in the limit.
```
cc @amirylm and @RyanRHall

amirylm replied:
will be handled in LM-37

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/opstack/l1_to_l2.go:
```diff
@@ -0,0 +1,564 @@
+		return nil, errors.New("no transfer nonces found in bridge payloads")
+	}
+
+	// TODO: reconfirm that all nonces should be the same across all payloads in a given round for OP
```
cc @makramkd you said this should hold true, but I wanted to reconfirm, otherwise this `QuorumizedBridgePayload()` function gets a little wonky since the `bridgeReturnData`/nonce should be unique for a given transfer

ogtownsend replied:
cc @amirylm

amirylm replied:
Yes seems that if upon mismatch we don't continue to look for a quorum of matching nonces then this function should be renamed..

ogtownsend replied:
LM-166

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/plugin.go:
```diff
@@ -218,16 +218,16 @@ func combinedUnexecutedTransfers(
-	for _, pendingTransfer := range pendingTransfers {
-		unexecuted = append(unexecuted, pendingTransfer)
-	}
+	unexecuted := make([]liquidityrebalancer.UnexecutedTransfer, 0, len(resolvedTransfersQuorum)+len(inflightTransfers)+len(pendingTransfers))
```
this was just a nit, felt like resolved --> inflight --> pending better represented the chronological ordering/stages that the transfers go through

ogtownsend commented on core/services/ocr2/plugins/liquiditymanager/bridge/opstack/l1_to_l2.go:
```diff
@@ -0,0 +1,569 @@
+		log.Fatalf("Failed to pack depositETHTo function call: %v", err)
+	}
+
+	// Estimate gas needed for the depositETHTo call issued from the L1 Bridge Adapter
```
update comment

amirylm commented on core/services/ocr2/plugins/liquiditymanager/bridge/opstack/l1_to_l2.go:
```diff
@@ -0,0 +1,570 @@
+	// TODO: maybe add check if this is a native transfer or ERC20 transfer
+	calldata, err := l1standardBridgeABI.Pack(
+		// If we're sending WETH, the bridge adapter unwraps it and calls depositETHTo on the native bridge
+		"depositETHTo",
```
extract const

amirylm commented on core/services/ocr2/plugins/liquiditymanager/bridge/arb/l1_to_l2.go:
```diff
@@ -86,17 +87,17 @@ func NewL1ToL2Bridge(
 	}
 
-	l1FilterName := fmt.Sprintf("ArbitrumL2ToL1Bridge-L1-LiquidityManager:%s-Local:%s-Remote:%s",
+	l1FilterName := fmt.Sprintf("ArbitrumL1ToL2Bridge-L1-LiquidityManager:%s-Local:%s-Remote:%s",
```
We can extract a function that accepts a name and applies this pattern.

ogtownsend committed:
PR feedback

ogtownsend committed:
Add Optimism L2 to L1 bridge interface offchain support (#959)

Similar to https://github.com/smartcontractkit/ccip/pull/840, this PR
adds the Optimism L2 to L1 bridge interface for the LiquidityManager.
Some things to note:
- Similar to the L1 to L2 flow, the Optimism L2 bridge adapter
increments a nonce with each transfer and each time `sendERC20` is
called. This nonce is included in multiple events throughout the
transfer's lifecycle and is how we match and link events together to
confirm they've reached the various states.
- The `GetTransfers()` function makes use of this nonce and implements
the matching logic to link these 3 events together and partition
transfers into categories:
- L2's LiquidityManager `LiquidityTransferred.bridgeReturnData`, which
should equal
- L1's LiquidityManager `FinalizationStepCompleted.bridgeSpecificData`
(once it's unwrapped a few times) which should equal
  - L1's LiquidityManager `LiquidityTransferred.bridgeSpecificData`
- See the `TODO (ogtownsend / amirylm): is the plugin able to handle the
case where we've already instructed the plugin to prove() ...` comment

amirylm commented on core/services/ocr2/plugins/liquiditymanager/bridge/bridge.go:
```diff
@@ -171,18 +173,71 @@ func (f *factory) initBridge(source, dest models.NetworkSelector) (Bridge, error
 		if source == models.NetworkSelector(chainsel.ETHEREUM_MAINNET.Selector) &&
-			dest != models.NetworkSelector(chainsel.ETHEREUM_MAINNET_ARBITRUM_1.Selector) {
-			return nil, fmt.Errorf("unsupported destination for eth mainnet l1 -> l2 bridge: %d, must be arb mainnet", dest)
+			dest != models.NetworkSelector(chainsel.ETHEREUM_MAINNET_ARBITRUM_1.Selector) &&
```
Let's use a static map of src to valid dest networks,
Otherwise it could be too complex when we scale to more networks.

e.g.:
```go
type Chains map[uint64][]uint64

func (c Chains) Supports(src, dest models.NetworkSelector) bool { 
// ... 
}

var chains = Chains{
  chainsel.ETHEREUM_MAINNET.Selector: []uint64{
    chainsel.ETHEREUM_MAINNET_ARBITRUM_1.Selector,
    chainsel.ETHEREUM_MAINNET_OPTIMISM_1.Selector,
  },
  // ...
}
```

ogtownsend committed:
Add supported bridge helper

ogtownsend committed:
Use TransferStatusReady status for OP L2 to L1 withdrawals that need to be proven

infiloop2 reviewed (approved):
accepting to unblock

RensR commented on .changeset/great-seals-rule.md:
```diff
@@ -0,0 +1,5 @@
+---
```
nit: we've stopped doing changesets in this repo. All of them can be removed and no new PR needs them

ogtownsend committed:
Remove changesets per Rens' comment
